### PR TITLE
Chrome DUT based on tgl_max98373_rt5682: override tplg and quirk data

### PR DESCRIPTION
**Pull request by sathya-nujella**
Chrome DUT based on tgl_max98373_rt5682: override tplg and quirk data

Below changes are done here:

1)Machine driver's platform data has a copy of acpi mach data.
So, override tplg info on this platform data and use this tplg info
while loading tplg binary. Updated sof pcm.c accordingly.

2) A Chrome System based on tgl_max98373_rt5682 has different SSP and tplg
configurations. Using DMI data of this variant DUT, overiride
quirk and tplg information.

### CODE REVIEW

**Comment by ranj063** [comment]
@dbaluta how would this change affect you?

File: sound/soc/sof/pcm.c [hunk 1]
```diff
@@ -756,6 +756,12 @@ static int sof_pcm_probe(struct snd_soc_component *component)
 	struct snd_sof_pdata *plat_data = sdev->pdata;
 	const char *tplg_filename;
 	int ret;
+	struct platform_device *pdev = plat_data->pdev_mach;
+	struct snd_soc_acpi_mach *mach = pdev->dev.platform_data;
```
Comment (lyakh) [inline sound/soc/sof/pcm.c]: I might be wrong but this doesn't look right to me. This is a generic pcm.c file, it has nothing to do with ACPI. And not every SOF implementation has ACPI information. Why not do that in `sof_acpi_probe()`? Any reason why that wouldn't work? And you want to use `dev_get_platdata();`

File: sound/soc/intel/boards/sof_rt5682.c [hunk 2]
```diff
@@ -66,12 +66,29 @@ struct sof_card_private {
...
 };
 
+const char *sof_override_tplg_name;
+
 static int sof_rt5682_quirk_cb(const struct dmi_system_id *id)
 {
 	sof_rt5682_quirk = (unsigned long)id->driver_data;
 	return 1;
 }
-
```
Comment (lyakh) [inline sound/soc/intel/boards/sof_rt5682.c]: please, keep empty lines between functions

File: sound/soc/intel/boards/sof_rt5682.c [hunk 3]
```diff
@@ -66,12 +66,29 @@ struct sof_card_private {
...
+static int sof_rt5682_tplg_cb(const struct dmi_system_id *id)
+{
+	sof_override_tplg_name = (const char *)id->driver_data;
+	return 1;
+}
```
Comment (lyakh) [inline sound/soc/intel/boards/sof_rt5682.c]: empty line

File: sound/soc/intel/boards/sof_rt5682.c [hunk 4]
```diff
@@ -66,12 +66,29 @@ struct sof_card_private {
...
+static const struct dmi_system_id sof_rt5682_tplg_table[] = {
+	{
+		.callback = sof_rt5682_tplg_cb,
+		.matches = {
+			DMI_MATCH(DMI_PRODUCT_FAMILY, "Google_Volteer"),
+			DMI_MATCH(DMI_PRODUCT_NAME, "Terrador"),
+		},
+		.driver_data = (void *)("sof-tgl-max98373-rt5682-ssp2.tplg"),
```
Comment (lyakh) [inline sound/soc/intel/boards/sof_rt5682.c]: I think I suggested before to remove the cast:
```
.driver_data = "sof-tgl-max98373-rt5682-ssp2.tplg",
```

File: sound/soc/sof/pcm.c (hunk 1 above)
Comment (ranj063) [reply sound/soc/sof/pcm.c]: @lyakh this change is not specific for ACPI platforms. For the Intel platforms, we always have this information in the machine driver platform device.  On the other hand, what I am not sure about is if this applies to DT platforms. Something for @dbaluta to confirm.

Comment (dbaluta) [reply sound/soc/sof/pcm.c]: @sathya-nujella @ranj063 this doesn't looks good from Device Tree perspective. If you want to override the tplg_filename in the worst case please create a generic snd_sof_* function that gets implemented by each platform.

This could apply to DT platforms also (overriding the name of the topology file). We use this kind of trick to update the name of the machine driver (will need to find a generic way, when have some time to upstream my patch).

https://github.com/dbaluta/linux/blob/next/sound/soc/sof/pcm.c#L793

To sum up, I'm not against this change if it is made in a generic way so that no ACPI stuff is added to generic pcm.c file.

Comment (plbossart) [reply sound/soc/sof/pcm.c]: We should do this in sof-pci-dev.c

We can add a new DMI table, to override specific topology files.

````c
if (dmi_check_system(google_alternate_topology_platforms)) {
		sof_pdata->tplg_filename = sof_override_tplg_name;
}
````
	
we might even be able to avoid this extra variable and directly get the data from the quirk.

Comment (sathya-nujella) [reply sound/soc/sof/pcm.c]: > We should do this in sof-pci-dev.c
> 
> We can add a new DMI table, to override specific topology files.
> 
> ```c
> if (dmi_check_system(google_alternate_topology_platforms)) {
> 		sof_pdata->tplg_filename = sof_override_tplg_name;
> }
> ```
> 
> we might even be able to avoid this extra variable and directly get the data from the quirk.

Thank you @plbossart . I have removed previous tplg dmi method and added in sof-pci-dev.

Comment (sathya-nujella) [reply sound/soc/sof/pcm.c]: > @sathya-nujella @ranj063 this doesn't looks good from Device Tree perspective. If you want to override the tplg_filename in the worst case please create a generic snd_sof_* function that gets implemented by each platform.
> 
> This could apply to DT platforms also (overriding the name of the topology file). We use this kind of trick to update the name of the machine driver (will need to find a generic way, when have some time to upstream my patch).
> 
> https://github.com/dbaluta/linux/blob/next/sound/soc/sof/pcm.c#L793
> 
> To sum up, I'm not against this change if it is made in a generic way so that no ACPI stuff is added to generic pcm.c file.

I removed earlier method and changed based on latest inputs. Thanks!

File: sound/soc/intel/boards/sof_rt5682.c (hunk 2 above)
Comment (sathya-nujella) [reply sound/soc/intel/boards/sof_rt5682.c]: > please, keep empty lines between functions

Done now.

File: sound/soc/intel/boards/sof_rt5682.c (hunk 3 above)
Comment (sathya-nujella) [reply sound/soc/intel/boards/sof_rt5682.c]: Done.

File: sound/soc/intel/boards/sof_rt5682.c (hunk 4 above)
Comment (sathya-nujella) [reply sound/soc/intel/boards/sof_rt5682.c]: sorry, I seem to have missed removing type casting. Now, I have updated it.

File: sound/soc/intel/boards/sof_rt5682.c (hunk 2 above)
Comment (lyakh) [reply sound/soc/intel/boards/sof_rt5682.c]: sorry, I also meant this empty line, looks like you're still removing it?

Comment (sathya-nujella) [reply sound/soc/intel/boards/sof_rt5682.c]: Thank you, corrected it now.

File: sound/soc/sof/sof-pci-dev.c [hunk 5]
```diff
@@ -35,8 +35,28 @@ static int sof_pci_debug;
...
+}
+
+static const struct dmi_system_id sof_tplg_table[] = {
+	{
+		.callback = sof_tplg_cb,
+		.matches = {
+			DMI_MATCH(DMI_PRODUCT_FAMILY, "Google_Volteer"),
+			DMI_MATCH(DMI_PRODUCT_NAME, "Terrador"),
+		},
+		.driver_data = ("sof-tgl-max98373-rt5682-ssp2.tplg"),
```
Comment (plbossart) [inline sound/soc/sof/sof-pci-dev.c]: I'd like the topology name to be less ambiguous, here we don't know if ssp2 applies to the amplifier or the codec.

Also IIRC we started with the headset codec first, then the amplifier, is this too lake to change the topology names for SoundWire TGL?

It'd make more sense to me to have rt5682-sspN-max98373-sspM, that we know exactly what goes where.
It'd be similar to the SoundWire topologies where we list the devices starting with the link0 which happens to be the headset.

Comment (sathya-nujella) [reply sound/soc/sof/sof-pci-dev.c]: 1. I have updated tplg name appending ssp id's next to amp names.  yes, this way no ambiguity in tplg names. Thank you.

2. on changing to soudnwire TGL topologies:  we only have one SKU as of now.. can we leave it for now and do when necessary arises and along with other changes?

**Comment by plbossart** [comment]
SOFCI TEST

**Comment by plbossart** [comment]
@sathya-nujella I was talking about a simple change like this:

````diff
diff --git a/sound/soc/intel/common/soc-acpi-intel-tgl-match.c b/sound/soc/intel/common/soc-acpi-intel-tgl-match.c
index 6816847bee40..355a6d8ac175 100644
--- a/sound/soc/intel/common/soc-acpi-intel-tgl-match.c
+++ b/sound/soc/intel/common/soc-acpi-intel-tgl-match.c
@@ -261,19 +261,19 @@ struct snd_soc_acpi_mach snd_soc_acpi_intel_tgl_machines[] = {
        },
        {
                .id = "10EC5682",
-               .drv_name = "tgl_max98357a_rt5682",
+               .drv_name = "tgl_rt5682_max98357a",
                .machine_quirk = snd_soc_acpi_codec_list,
                .quirk_data = &tgl_codecs,
                .sof_fw_filename = "sof-tgl.ri",
-               .sof_tplg_filename = "sof-tgl-max98357a-rt5682.tplg",
+               .sof_tplg_filename = "sof-tgl-rt5682-max98357a.tplg",
        },
        {
                .id = "10EC5682",
-               .drv_name = "tgl_max98373_rt5682",
+               .drv_name = "tgl_rt5682_max98373",
                .machine_quirk = snd_soc_acpi_codec_list,
                .quirk_data = &tgl_max98373_amp,
                .sof_fw_filename = "sof-tgl.ri",
-               .sof_tplg_filename = "sof-tgl-max98373-rt5682.tplg",
+               .sof_tplg_filename = "sof-tgl-rt5682-max98373.tplg",
        },
        {},
 };
@@ -319,7 +319,7 @@ struct snd_soc_acpi_mach snd_soc_acpi_intel_tgl_sdw_machines[] = {
                .links = tgl_chromebook_base,
                .drv_name = "sof_sdw",
                .sof_fw_filename = "sof-tgl.ri",
-               .sof_tplg_filename = "sof-tgl-sdw-max98373-rt5682.tplg",
+               .sof_tplg_filename = "sof-tgl-sdw-rt5682-max98373.tplg",
        },
        {
                .link_mask = 0x1, /* this will only enable rt5682 for now */
````

**Comment by sathya-nujella** [comment]
> LGTM, but can you also send a PR to change the topology names for rt5682-max98373 in sound/soc/intel/common and also in SOF tools/topology/CMakefile
> 
> Thanks @sathya-nujella

sure  @plbossart , I will do.

**Comment by sathya-nujella** [comment]
> @sathya-nujella I was talking about a simple change like this:
> 
> ```diff
> diff --git a/sound/soc/intel/common/soc-acpi-intel-tgl-match.c b/sound/soc/intel/common/soc-acpi-intel-tgl-match.c
> index 6816847bee40..355a6d8ac175 100644
> --- a/sound/soc/intel/common/soc-acpi-intel-tgl-match.c
> +++ b/sound/soc/intel/common/soc-acpi-intel-tgl-match.c
> @@ -261,19 +261,19 @@ struct snd_soc_acpi_mach snd_soc_acpi_intel_tgl_machines[] = {
>         },
>         {
>                 .id = "10EC5682",
> -               .drv_name = "tgl_max98357a_rt5682",
> +               .drv_name = "tgl_rt5682_max98357a",
>                 .machine_quirk = snd_soc_acpi_codec_list,
>                 .quirk_data = &tgl_codecs,
>                 .sof_fw_filename = "sof-tgl.ri",
> -               .sof_tplg_filename = "sof-tgl-max98357a-rt5682.tplg",
> +               .sof_tplg_filename = "sof-tgl-rt5682-max98357a.tplg",
>         },
>         {
>                 .id = "10EC5682",
> -               .drv_name = "tgl_max98373_rt5682",
> +               .drv_name = "tgl_rt5682_max98373",
>                 .machine_quirk = snd_soc_acpi_codec_list,
>                 .quirk_data = &tgl_max98373_amp,
>                 .sof_fw_filename = "sof-tgl.ri",
> -               .sof_tplg_filename = "sof-tgl-max98373-rt5682.tplg",
> +               .sof_tplg_filename = "sof-tgl-rt5682-max98373.tplg",
>         },
>         {},
>  };
> @@ -319,7 +319,7 @@ struct snd_soc_acpi_mach snd_soc_acpi_intel_tgl_sdw_machines[] = {
>                 .links = tgl_chromebook_base,
>                 .drv_name = "sof_sdw",
>                 .sof_fw_filename = "sof-tgl.ri",
> -               .sof_tplg_filename = "sof-tgl-sdw-max98373-rt5682.tplg",
> +               .sof_tplg_filename = "sof-tgl-sdw-rt5682-max98373.tplg",
>         },
>         {
>                 .link_mask = 0x1, /* this will only enable rt5682 for now */
> ```

Sure @plbossart , understood. I will do in a separate PR. Thanks!

File: sound/soc/sof/sof-pci-dev.c [hunk 6]
```diff
@@ -35,8 +35,28 @@ static int sof_pci_debug;
...
+}
+
+static const struct dmi_system_id sof_tplg_table[] = {
+	{
+		.callback = sof_tplg_cb,
+		.matches = {
+			DMI_MATCH(DMI_PRODUCT_FAMILY, "Google_Volteer"),
+			DMI_MATCH(DMI_PRODUCT_NAME, "Terrador"),
+		},
+		.driver_data = ("sof-tgl-rt5682-ssp0-max98373-ssp2.tplg"),
```
Comment (lyakh) [inline sound/soc/sof/sof-pci-dev.c]: I still don't understand why a simple
```
.driver_data = "sof-tgl-rt5682-ssp0-max98373-ssp2.tplg",
```
isn't enough, sorry.

Comment (sathya-nujella) [reply sound/soc/sof/sof-pci-dev.c]: I have changed based on this input.

**Comment by sathya-nujella** [comment]
rebased on top of latest sof-dev

File: sound/soc/sof/sof-pci-dev.c [hunk 7]
```diff
@@ -35,8 +35,28 @@ static int sof_pci_debug;
 module_param_named(sof_pci_debug, sof_pci_debug, int, 0444);
 MODULE_PARM_DESC(sof_pci_debug, "SOF PCI debug options (0x0 all off)");
 
+const char *sof_override_tplg_name;
```
Comment (lyakh) [inline sound/soc/sof/sof-pci-dev.c]: Oh, sorry for missing this during the first review rounds, I was thinking about getting rid of this global variable and missed to check it properly: this must be `static` of course.

File: sound/soc/sof/sof-pci-dev.c [hunk 8]
```diff
@@ -35,8 +35,28 @@ static int sof_pci_debug;
...
+
 #define SOF_PCI_DISABLE_PM_RUNTIME BIT(0)
 
+static int sof_tplg_cb(const struct dmi_system_id *id)
+{
+	sof_override_tplg_name = (const char *)id->driver_data;
```
Comment (lyakh) [inline sound/soc/sof/sof-pci-dev.c]: and you don't need this type-cast either.

File: sound/soc/sof/sof-pci-dev.c (hunk 7 above)
Comment (sathya-nujella) [reply sound/soc/sof/sof-pci-dev.c]: np, thank you. I made it static now and updated patch.

File: sound/soc/sof/sof-pci-dev.c (hunk 8 above)
Comment (sathya-nujella) [reply sound/soc/sof/sof-pci-dev.c]: Done.

**Comment by sathya-nujella** [comment]
> sorry, I should have noticed these before :-(

np thank you for inputs.

File: sound/soc/intel/boards/sof_rt5682.c [hunk 9]
```diff
@@ -119,6 +119,19 @@ static const struct dmi_system_id sof_rt5682_quirk_table[] = {
 		.driver_data = (void *)(SOF_RT5682_MCLK_EN |
 					SOF_RT5682_SSP_CODEC(0)),
 	},
+	{
```
Comment (ranj063) [inline sound/soc/intel/boards/sof_rt5682.c:122]: @sathya-nujella there's a typo in the commit message "overiride"

Comment (sathya-nujella) [reply sound/soc/intel/boards/sof_rt5682.c:122]: oh, sorry for that typo. fixed it now.
